ruqqus/ruqqus
# Alembic Integration

Implementation of [Alembic ](https://alembic.sqlalchemy.org/en/latest/tutorial.html) database change management as a replacement for the Schema.txt file.

This assume starting with an empty database as a starting point. Happy to provide help on integrating it into the production db if desired.

**TODO:** Will need to create some wiki pages covering the setup process in more detail, likely replacing whats in the README.

### Notes
Took some notes as I came across things that I thought were weird or could be problems.
#### Duplicate Indexes:
Found the following duplicate indexes:
##### comments
    "comments_parent_id_idx" btree (parent_comment_id)
    
##### contributors
    "contributors_board_index" btree (board_id)

##### submissions
    "domain_ref_idx" btree (domain_ref)
    "post_18_index" btree (over_18)
    "post_author_index" btree (author_id)
    "post_offensive_index" btree (is_offensive)

##### subscriptions
    "sub_user_index" btree (user_id)

#### Records:
* `score_top` is an Integer, all other score fields are Float. Unsure if this is intended?
* `rules(rule_html)` unbounded charvar
* `users(real_id)` unbounded charvar

#### Misc:
* `commentvotes` had no primary key
* `comments_aux` is a poor table name, indexes (among other things) are easily confused with `comments`. Prefixed indexes as `commentsaux` because of this
* `submissions_aux` same note as `comments_aux`

## Description
1. Run pip to get the latest requirements
2. Ensure PYTHONPATH envar is set to point at ruqqus
3. Edit `alembic.ini` and verify `sqlalchemy.url` points to the correct database.

Once complete you can run the following commands:

`alembic upgrade head` to initialize an empty database schema
`alembic -x data=true upgrade head` to initialize the database schema and populate it with the user ruqqie, the +general guild, and badge_defs data (just the beta badge atm)

`alembic downgrade -1` will reset back to an empty database

`alembic revision -m "description"` generates a new revision in the versions folder for you to populate with changes. Run `alembic upgrade head` to apply it.

Also have 3 semi-related fixes:

1. Fixed a bug when no data populates the badwords database
2. The cors check now allows http if env var SESSION_COOKIE_SECURE is set as false (allows http dev locally)
3. Admins can always create a guild regardless of rep count

## Motivation and Context
Now have the ability to track database changes going forward. Developers can also now easily submit changes to the database.

## How Has This Been Tested?
First ran this script against one database and imported the Schema.txt file into another.

Table by table compared the table descriptions (\d tablename)

After verifying tables match, ran ruqqus locally and logged in, posted in +general, created a guild, created a user, and commented.

## Screenshots (if appropriate):

## Types of changes


## Checklist:
